File: easy-projects/easy-common/src/main/java/xyz/srclab/common/util/proxy/ProxyMethod.java
package xyz.srclab.common.util.proxy;

import xyz.srclab.common.invoke.MethodInvoker;

import java.lang.reflect.Method;

/**
 * @author sunqian
 */
public interface ProxyMethod {

    Object invoke(Object object, Object[] args, Method method, MethodInvoker superInvoker);
}

File: boat-core/src/main/kotlin/xyz/srclab/common/proxy/ProxyMethod.kt
package xyz.srclab.common.proxy

import xyz.srclab.common.invoke.StaticInvoke
import java.lang.reflect.Method

/**
 * To describe proxy method info.
 */
interface ProxyMethod {

    /**
     * Returns whether this [ProxyMethod] is valid for given [method].
     */
    fun isProxy(method: Method): Boolean

    /**
     * Proxy action.
     */
    operator fun invoke(
        sourceObject: Any,
        sourceMethod: Method,
        sourceInvoke: StaticInvoke,
        args: Array<out Any?>?
    ): Any?
}